calculation/sha_calc/sha_calc/directivity/bea20/distributions.py

```diff
@@ -0,0 +1,314 @@
+    upper, lower = distribution.cdf((1, 0))
+    dist_range = upper - lower
+    truncated_points = (np.random.uniform(0, 1, hypo_down_dip)) * dist_range + lower
+    down_dip_distribution = distribution.ppf(truncated_points)
+
+    planes_list = []
+    planes_index = []
+
+    for distance in distances:
+        current_length = 0
+        planes_copy = [plane.copy() for plane in planes]
+        for index, plane in enumerate(planes_copy):
+            if current_length < distance < (current_length + plane["length"]):
+                for depth in down_dip_distribution:
+                    planes_depth_copy = [plane.copy() for plane in planes]
+                    planes_depth_copy[index]["shyp"] = distance - (total_length / 2)
+                    planes_depth_copy[index]["dhyp"] = plane["width"] * depth
+                    planes_index.append(index)
+                    planes_list.append(planes_depth_copy)
+            current_length += plane["length"]
+
+    return planes_list, planes_index, []
+
+
+def even_grid(
+    hypo_along_strike: int,
+    hypo_down_dip: int,
+    planes: List,
+    event_type: EventType,
+    total_length: float,
+):
+    """
+    Uses an evenly spaced grid across strike and dip based on the amount of hypocentres specified across both lengths
+
+    Parameters
+    ----------
+    hypo_along_strike: int
+        Number of hypocentres across strike
+    hypo_down_dip: int
+        Number of hypocentres down dip
+    planes: list
+        The planes to adjust and set the hypocentre on
+    event_type: EventType
+        The event type Strike_slip, dip_slip or all for determining the down dip distribution function
+    total_length: float
+        The total length of the fault
+    """
+
+    gap = total_length / (hypo_along_strike + 1)
+    distances = [gap * (x + 1) for x in range(hypo_along_strike)]
+
+    down_gap = planes[0]["width"] / (hypo_down_dip + 1)
+    down_dip_values = [down_gap * (x + 1) for x in range(hypo_down_dip)]
+
+    # Works out the distances across strike of the fault for each hypocentre
+    # Based on a normal distribution and truncated between 0 and 1
+    mu, sigma = 0.5, 0.23
+    lower, upper = 0, 1
+    strike_distribution = stats.truncnorm(
+        (lower - mu) / sigma, (upper - mu) / sigma, loc=mu, scale=sigma
+    )
+    strike_weight_dist = strike_distribution.pdf(
+        [x / total_length for x in distances]
+    ) * (1 / hypo_along_strike)
+
+    # Works out the depth method for down dip placement of hypocentres
+    # Based on Weilbull or Gamma distributions depending on the EventType
+    if event_type == EventType.DIP_SLIP:
+        down_dip_distribution = gamma_truncated(a=0, b=1)
+    elif event_type == EventType.STRIKE_SLIP:
+        down_dip_distribution = weibull_truncated_strike_slip(a=0, b=1)
+    else:
+        down_dip_distribution = weibull_truncated_oblique(a=0, b=1)
+    dip_weight_dist = down_dip_distribution.pdf(
```

Comment:
The following few lines can be simplified with numpy. In general I would suggest using numpy arrays for calculations instead of for loops. When working with arrays of numbers numpy arrays >> lists.
`dip_weight_dist = down_dip_distribution.pdf(np.asarray(down_dip_values) / planes[0]["width"]) * (1 / hypo_down_dip)`

`combo_weight_dist = (strike_weight_dist[:, np.newaxis] * dip_weight_dist).ravel()`

And then to normalise its just
`combo_weight_dist = combo_weight_dist / combo_weight_dist.sum()`

Also I suggest removing the `dist` suffix from these variables (assuming dist stands for distribution) as these are weights from a distribution, but they are not a distribution, i.e. they are just weights.

There are more statements further up (197-212) that can be simplified with numpy

Reply:
One other thing, I suggest adding an assert statement that ensures that the width of all planes is the same, just a nice sanity check. 
In general I think its a good idea to add assert statements when making assumptions, exceptions are way better than silent incorrect calculations.